graph/preprocessor.go

```diff
@@ -0,0 +1,211 @@
+	remoteClient := http.Client{
+		Timeout: time.Second * 2, // Maximum of 2 secs
+	}
+
+	req, err := http.NewRequest(http.MethodGet, url, nil)
+	if err != nil {
+		return err
+	}
+
+	res, getErr := remoteClient.Do(req)
+	if getErr != nil {
+		return getErr
+	}
+
+	data, readErr := ioutil.ReadAll(res.Body)
+	if readErr != nil {
+		return readErr
+	}
+
+	return readAliasFromBytes(data, preTask)
+}
+
+/* Parses out local alias files and adds their content to the passed in
+PreTask. Note alias definitions already in preTask will not be
+overwritten. */
+func addAliasFromFile(preTask *PreTask, fileURI string) error {
+
+	data, fileReadingError := ioutil.ReadFile(fileURI)
+	if fileReadingError != nil {
+		return fileReadingError
+	}
+	return readAliasFromBytes(data, preTask)
+}
+
+/* Parses out alias  definitions from a given bytes array and appends
+them to the PreTask. Note alias definitions already in preTask will
+not be overwritten even if present in the array. */
+func readAliasFromBytes(data []byte, preTask *PreTask) error {
+
+	aliasMap := &map[string]string{}
+
+	if err := yaml.Unmarshal(data, aliasMap); err != nil {
+		return err
+	}
+
+	for key, value := range *aliasMap {
+		if _, ok := preTask.AliasMap[key]; !ok {
+			preTask.AliasMap[key] = value
+		}
+	}
+	return nil
+}
+
+// PreprocessString handles managing alias definitions from a provided string definitions expected to be in JSON format.
+func PreprocessString(preTask *PreTask, str string) (string, error) {
+	//preTask.loadGlobalDefinitions TODO?
+
+	// Load Remote/Local alias definitions
+	if externalDefinitionErr := preTask.loadExternalAlias(); externalDefinitionErr != nil {
+		return "", externalDefinitionErr
+	}
+	//Validate alias definitions
+	if improperFormatErr := preTask.resolveMapAndValidate(); improperFormatErr != nil {
+		return "", improperFormatErr
+	}
+	var out strings.Builder
+	var command strings.Builder
+	ongoingCmd := false
+
+	// Search and Replace
+	for _, char := range str {
+		if ongoingCmd {
+			//Maybe just checking if non alphanumeric, only allow alpha numeric aliases?
+			if strings.Contains(")}/ .,;]&|'~\n\t", string(char)) { // Delineates the end of an alias
+				resolvedCommand, commandPresent := preTask.AliasMap[command.String()]
+				if !commandPresent {
+					return "", errUnknownAlias
+				}
+
+				out.WriteString(resolvedCommand)
+				if char != preTask.directive {
+					ongoingCmd = false
+					out.WriteRune(char)
+				}
+				command.Reset()
+
+			} else {
+				command.WriteRune(char)
+			}
+		} else if char == preTask.directive {
+
+			if ongoingCmd { // Escape character triggered
+				out.WriteRune(preTask.directive)
+				ongoingCmd = false
+				continue
+			}
+
+			ongoingCmd = true
+			continue
+		} else {
+			out.WriteRune(char)
+		}
+	}
+
+	return out.String(), nil
+}
+
+// PreprocessBytes Handles files or byte encoded data that can be parsed through pre processing
+func PreprocessBytes(data []byte) ([]byte, error) {
+	preTask := &PreTask{}
```

Comment:
an optimization here - if there are no aliases defined then there's no reason to replace anything or unmarshal it twice - since the end result will be the same. So it might make sense to just unmarshal the Task as a Task normally.

Reply:
That is true and I've given it some thought, nonetheless I am not sure if there is a way to prevent the double unmarshaling as I need to first confirm there are no aliases defined. The alternative would be to unmarshall the whole thing (as in no PreTask struct and just an extended task struct) and then modifying this resulting structure but that would require additional parsing of the resulting dynamic map and I am not sure that's much better. Do you have any thoughts about how I may approach this?